File: Tiny Types/Java/Cpf.java
/*
 * Copyright (c) 2014 HollowSoft @IgorMorais
 *
 * Licensed under the Apache License, Version 2.0 (the “License”);
 * you may not use this file except in compliance with the License.
 *
 *          You may obtain a copy of the License at
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an “AS IS” BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package tinytype;

/**
 * @author Igor Morais
 * @author mor41s.1gor@gmail.com
 */
public final class Cpf {

    /**
     * Position of the first dot.
     */
    private static final byte POSITION_DOT_ONE = 3;

    /**
     * Position of the second dot.
     */
    private static final byte POSITION_DOT_TWO = 7;

    /**
     * Position of the hyphen.
     */
    private static final byte POSITION_HYPHEN = 11;

    /**
     * Length of the {@code Cpf} number without masks.
     */
    private static final byte LENGTH_UNFORMATTED = 11;

    /**
     * Length of the {@code Cpf} number with masks.
     */
    private static final byte LENGTH_FORMATTED = 14;

    /**
     * String empty value.
     */
    private static final String EMPTY = "";

    /**
     * Masks.
     */
    private static final char DOT = '.';
    private static final char HYPHEN = '-';
    private static final String MASK = "[.-]";

    /**
     * The {@code Cpf} number.
     */
    private String number;

    /**
     * Default public constructor.
     */
    public Cpf() {

    }

    /**
     *
     * @param number
     *            The {@code Cpf} number.
     */
    public Cpf(final String number) {
        setNumber(number);
    }

    /**
     *
     * @param number
     * @return {@code True} if the {@code String} have only numbers, otherwise
     *         return {@code false}.
     */
    private static boolean hasOnlyNumbers(final String number) {
        boolean onlyNumbers = true;

        try {

            Long.parseLong(number);

        } catch (final NumberFormatException e) {
            onlyNumbers = false;
        }

        return onlyNumbers;
    }

    /**
     *
     * @param sum
     * @return
     */
    private static int verify(final int sum) {
        final int value = sum * 10 % 11;

        return value == 10 || value == 11 ? 0 : value;
    }

    /**
     *
     * @param number
     * @return {@code True} if the {@code Cpf} number is valid, otherwise return
     *         {@code false}.
     */
    private static boolean hasValidNumber(final String number) {
        boolean validNumber = true;

        byte amount = 0;

        final char[] numberArray = number.replaceAll(MASK, EMPTY).toCharArray();

        for (byte i = 1; i < numberArray.length; i++) {

            if (numberArray[0] == numberArray[i]) {
                amount++;
            }
        }

        if (amount == 10) {
            validNumber = false;
        }

        return validNumber;
    }

    /**
     *
     * @param number
     * @return
     */
    public static String format(final String number) {
        if (number == null || number.isEmpty()) {
            throw new IllegalArgumentException("The number cannot be nul or empty.");
        }

        return new StringBuilder(number).insert(POSITION_DOT_ONE, DOT).insert(POSITION_DOT_TWO, DOT)
                .insert(POSITION_HYPHEN, HYPHEN).toString();
    }

    /**
     *
     * @param number
     * @return
     */
    public static String unformat(final String number) {
        if (number == null || number.isEmpty()) {
            throw new IllegalArgumentException("The number cannot be nul or empty.");
        }

        return number.replaceAll(MASK, EMPTY);
    }

    /**
     *
     * @param number
     * @return {@code True} if the {@code Cpf} number have a valid format,
     *         otherwise return {@code false}.
     */
    public static boolean hasValidFormat(final String number) {
        if (number == null || number.isEmpty()) {
            throw new IllegalArgumentException("The number cannot be nul or empty.");
        }

        boolean validFormat = false;

        if (number.length() == LENGTH_UNFORMATTED) {

            if (hasOnlyNumbers(number)) {
                validFormat = true;
            }

        } else if (number.length() == LENGTH_FORMATTED) {

            if (number.charAt(POSITION_DOT_ONE) == DOT && number.charAt(POSITION_DOT_TWO) == DOT
                    && number.charAt(POSITION_HYPHEN) == HYPHEN) {

                if (hasOnlyNumbers(number.replaceAll(MASK, EMPTY))) {
                    validFormat = true;
                }
            }
        }

        return validFormat;
    }

    /**
     *
     * @param number
     * @return {@code True} if the {@code Cpf} number is valid, otherwise return
     *         {@code false}.
     */
    public static boolean validate(final String number) {
        if (number == null || number.isEmpty()) {
            throw new IllegalArgumentException("The number cannot be nul or empty.");
        }

        boolean valid = false;

        if (hasValidFormat(number) && hasValidNumber(number)) {

            final char[] numberArray = number.length() == LENGTH_FORMATTED ? number.replaceAll(MASK, EMPTY)
                    .toCharArray() : number.toCharArray();

            int sum = 0;

            int verificationDigit = Character.getNumericValue(numberArray[9]);

            for (byte i = 0; i < numberArray.length - 2; i++) {

                sum += Character.getNumericValue(numberArray[i]) * (numberArray.length - 1 - i);
            }

            if (verify(sum) == verificationDigit) {

                sum = 0;

                verificationDigit = Character.getNumericValue(numberArray[10]);

                for (byte i = 0; i < numberArray.length - 1; i++) {

                    sum += Character.getNumericValue(numberArray[i]) * (numberArray.length - i);
                }

                if (verify(sum) == verificationDigit) {
                    valid = true;
                }
            }
        }

        return valid;
    }

    /**
     *
     * @return The unformatted {@code Cpf} number.
     */
    public String getUnformattedNumber() {
        return number;
    }

    /**
     *
     * @return The formatted {@code Cpf} number.
     */
    public String getFormattedNumber() {
        return new StringBuilder(number).insert(POSITION_DOT_ONE, DOT).insert(POSITION_DOT_TWO, DOT)
                .insert(POSITION_HYPHEN, HYPHEN).toString();
    }

    /**
     *
     * @param number
     *            The {@code Cpf} number.
     */
    public void setNumber(final String number) {
        if (number == null || number.isEmpty()) {
            throw new IllegalArgumentException("The number cannot be nul or empty.");
        }

        this.number = number.replaceAll(MASK, EMPTY);
    }

    /**
     *
     * @return {@code True} if the {@code Cpf} number is valid, otherwise return
     *         {@code false}.
     */
    public boolean isValid() {
        return validate(number);
    }
}

File: Tiny Types/C#/Cpf.cs
/*
 * Copyright (c) 2014 HollowSoft @IgorMorais
 *
 * Licensed under the Apache License, Version 2.0 (the “License”);
 * you may not use this file except in compliance with the License.
 *
 *          You may obtain a copy of the License at
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an “AS IS” BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

using System;
using System.Linq;
using System.Text;

namespace TinyType
{
    /// <summary>
    /// Igor Morais
    /// mor41s.1gor@gmail.com
    /// </summary>
    public sealed class Cpf
    {
        /// <summary>
        /// Position of the first dot.
        /// </summary>
        private const byte PositionDotOne = 3;

        /// <summary>
        /// Position of the second dot.
        /// </summary>
        private const byte PositionDotTwo = 7;

        /// <summary>
        /// Position of the hyphen.
        /// </summary>
        private const byte PositionHyphen = 11;

        /// <summary>
        /// Length of the Cpf number without masks.
        /// </summary>
        private const byte LengthUnformatted = 11;

        /// <summary>
        /// Length of the Cpf number with masks.
        /// </summary>
        private const byte LengthFormatted = 14;

        /// <summary>
        /// Masks.
        /// </summary>
        private const char Dot = '.';
        private const char Hyphen = '-';
        private const string Mask = "[.-]";

        /// <summary>
        /// The Cpf number.
        /// </summary>
        private string Number;

        /// <summary>
        /// Default public constructor.
        /// </summary>
        public Cpf()
        {

        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="number">The Cpf number.</param>
        public Cpf(string number)
        {
            SetNumber(number);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="number"></param>
        /// <returns>True if the string have only numbers, otherwise return false.</returns>
        private static bool HasOnlyNumbers(string number)
        {
            bool OnlyNumbers = true;

            try
            {
                long.Parse(number);
            }
            catch (FormatException)
            {
                OnlyNumbers = false;
            }

            return OnlyNumbers;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="sum"></param>
        /// <returns></returns>
        private static int Verify(int sum)
        {
            int Value = sum * 10 % 11;

            return Value == 10 || Value == 11 ? 0 : Value;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="number"></param>
        /// <returns>True if the Cpf number is valid, otherwise return false.</returns>
        private static bool HasValidNumber(string number)
        {
            bool ValidNumber = true;

            byte Amount = 0;

            char[] NumberArray = number.Replace(Mask, string.Empty).ToCharArray();

            for (byte i = 1; i < NumberArray.Length; i++)
            {
                if (NumberArray[0] == NumberArray[i])
                {
                    Amount++;
                }
            }

            if (Amount == 10)
            {
                ValidNumber = false;
            }

            return ValidNumber;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="number"></param>
        /// <returns></returns>
        public static string Format(string number)
        {
            if (string.IsNullOrEmpty(number))
            {
                throw new ArgumentException("The Number cannot be null or empty.");
            }

            return new StringBuilder(number)
                .Insert(PositionDotOne, Dot)
                .Insert(PositionDotTwo, Dot)
                .Insert(PositionHyphen, Hyphen)
                .ToString();
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="number"></param>
        /// <returns></returns>
        public static string Unformat(string number)
        {
            if (string.IsNullOrEmpty(number))
            {
                throw new ArgumentException("The Number cannot be null or empty.");
            }

            return number.Replace(Mask, string.Empty);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="number"></param>
        /// <returns>True if the Cpf number have a valid format, otherwise return false.</returns>
        public static bool HasValidFormat(string number)
        {
            if (string.IsNullOrEmpty(number))
            {
                throw new ArgumentException("The Number cannot be null or empty.");
            }

            bool ValidFormat = false;

            if (number.Length == LengthUnformatted)
            {
                if (HasOnlyNumbers(number))
                {
                    ValidFormat = true;
                }
            }
            else if (number.Length == LengthFormatted)
            {
                if (number.ElementAt(PositionDotOne) == Dot &&
                    number.ElementAt(PositionDotTwo) == Dot &&
                    number.ElementAt(PositionHyphen) == Hyphen)
                {
                    if (HasOnlyNumbers(number.Replace(Mask, string.Empty)))
                    {
                        ValidFormat = true;
                    }
                }
            }

            return ValidFormat;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="number"></param>
        /// <returns>True if the Cpf number is valid, otherwise return false.</returns>
        public static bool Validate(string number)
        {
            if (string.IsNullOrEmpty(number))
            {
                throw new ArgumentException("The Number cannot be null or empty.");
            }

            bool Valid = false;

            if (HasValidFormat(number) && HasValidNumber(number))
            {
                char[] NumberArray = number.Length == LengthFormatted ?
                    number.Replace(Mask, String.Empty).ToCharArray() : number.ToCharArray();

                int Sum = 0;

                int VerificationDigit = (int)char.GetNumericValue(NumberArray[9]);

                for (byte i = 0; i < NumberArray.Length - 2; i++)
                {
                    Sum += (int)char.GetNumericValue(NumberArray[i]) * (NumberArray.Length - 1 - i);
                }

                if (Verify(Sum) == VerificationDigit)
                {
                    Sum = 0;

                    VerificationDigit = (int)char.GetNumericValue(NumberArray[10]);

                    for (byte i = 0; i < NumberArray.Length - 1; i++)
                    {
                        Sum += (int)char.GetNumericValue(NumberArray[i]) * (NumberArray.Length - i);
                    }
                }

                if (Verify(Sum) == VerificationDigit)
                {
                    Valid = true;
                }
            }

            return Valid;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns>The unformatted Cpf number.</returns>
        public string GetUnformattedNumber()
        {
            return Number;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns>The formatted Cpf number.</returns>
        public string GetFormattedNumber()
        {
            return new StringBuilder(Number)
                .Insert(PositionDotOne, Dot)
                .Insert(PositionDotTwo, Dot)
                .Insert(PositionHyphen, Hyphen)
                .ToString();
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="number">The Cpf number.</param>
        public void SetNumber(string number)
        {
            if (string.IsNullOrEmpty(number))
            {
                throw new ArgumentException("The Number cannot be null or empty.");
            }

            this.Number = number.Replace(Mask, String.Empty);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns>True if the Cpf number is valid, otherwise return false.</returns>
        public bool IsValid()
        {
            return Validate(Number);
        }
    }
}
